js-packages/@prestojs/viewmodel/src/fields/AsyncChoices.ts

```diff
@@ -0,0 +1,242 @@
+import { getId, getRichLabel, isIdentifiable, isLabeled } from '@prestojs/util';
+
+/**
+ * A single choice
+ */
+export interface Choice<T> {
+    value: T;
+    label: React.ReactNode;
+    [x: string]: any;
+}
+/**
+ * Grouped choices - a label for the grouping and an array of choices for that label
+ */
+export type ChoicesGrouped<T> = [string, Choice<T>[]];
+
+/**
+ * Interface for asynchronous choices.
+ *
+ * A choice is a `value` that identifies the item (eg. an id) and a `label` that
+ * describes the item and is shown to users (eg. the name).
+ *
+ * When `multiple` is true multiple values can be selected.
+ *
+ * To define async choices two things are required:
+ *
+ * 1) A function to resolve existing value(s). This is used when viewing existing
+ *    value(s) and need label(s) to show (eg. when displaying a choice on a detail
+ *    view or rendering the value selected on a select widget).
+ * 2) A function to list & filter the available  choices. This is used when selecting
+ *    a value (eg. the options shown in a select widget).
+ *
+ * Both of these functions may need to store state (eg. pagination for a listing)
+ * or access things from context (eg. read values from a cache). This can be done
+ * via two hooks - `useListProps` and `useRetrieveProps`. This functions should be
+ * called from a component or hook that deals with async choices when calling
+ * `list` and `retrieve` respectively. The return value from the hook is passed to
+ * the corresponding function.
+ *
+ * For a default implementation see [AsyncChoices](doc:AsyncChoices)
+ *
+ * @extract-docs
+ * @menu-group Async Choices
+ */
+export interface AsyncChoicesInterface<ItemType, ValueType> {
+    /**
+     * If true then multiple values can be selected. When this is true
+     * retrieve() will be passed and return an array rather than a single value.
+     */
+    multiple: boolean;
+    /**
+     * Hook that returns dependencies for listing choices. This is useful to store state
+     * for things like pagination. The value returned here is passed to `list`.
+     *
+     * What this function is passed depends on the implementation but when used with
+     * [useAsyncChoices](doc:useAsyncChoices) it will be passed `value` (the currently
+     * selected value), `query` (the query object eg. to filter results with) and `listOptions`
+     * (any additional options passed on the `listOptions` prop to `useAsyncChoices`).
+     * Official presto widgets all use `useAsyncChoices` and so will use these parameters.
+     */
+    useListProps(args: any): any;
```

Comment:
Can you give an example here of what a "dependency" is? How does it differ from the actual data itself?

eg if I have a list of organisations & their countries then is this supposed to query the country list or is that supposed to be a join that is going as part of the data returned by `list()` ?

Reply:
oops I changed the name from `useListDeps` to `useListProps` but still use dependencies in the comment.

The props returned here is anything required by the eg. widget or hooks that use the async choices. So pagination is the common thing:

```ts
useListProps() {
  const paginator = usePaginator(endpoints.list);
  return { paginator };
}
```

It also can be used to transform the props passed in - so if you wanted to change how the keywords from a generic widget are transformed before they are passed through to the async function (eg. the endpoint) you can do this here without having to have a custom widget.